src/StorageRent.sol

```diff
@@ -0,0 +1,578 @@
+        uint256 _initialDeprecationPeriod,
+        uint256 _initialUsdUnitPrice,
+        uint256 _initialMaxUnits,
+        uint256 _initialPriceFeedCacheDuration,
+        uint256 _initialUptimeFeedGracePeriod,
+        address _initialVault,
+        address _initialAdmin,
+        address _initialOperator,
+        address _initialTreasurer
+    ) {
+        priceFeed = _priceFeed;
+        uptimeFeed = _uptimeFeed;
+
+        deprecationTimestamp = block.timestamp + _initialDeprecationPeriod;
+        emit SetDeprecationTimestamp(0, deprecationTimestamp);
+
+        usdUnitPrice = _initialUsdUnitPrice;
+        emit SetPrice(0, _initialUsdUnitPrice);
+
+        maxUnits = _initialMaxUnits;
+        emit SetMaxUnits(0, _initialMaxUnits);
+
+        priceFeedCacheDuration = _initialPriceFeedCacheDuration;
+        emit SetCacheDuration(0, _initialPriceFeedCacheDuration);
+
+        uptimeFeedGracePeriod = _initialUptimeFeedGracePeriod;
+        emit SetGracePeriod(0, _initialUptimeFeedGracePeriod);
+
+        vault = _initialVault;
+        emit SetVault(address(0), _initialVault);
+
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(ADMIN_ROLE, _initialAdmin);
+        _grantRole(OPERATOR_ROLE, _initialOperator);
+        _grantRole(TREASURER_ROLE, _initialTreasurer);
+
+        _refreshPrice();
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                               MODIFIERS
+    //////////////////////////////////////////////////////////////*/
+
+    modifier whenNotDeprecated() {
+        if (block.timestamp >= deprecationTimestamp) revert ContractDeprecated();
+        _;
+    }
+
+    modifier onlyAdmin() {
+        if (!hasRole(ADMIN_ROLE, msg.sender)) revert NotAdmin();
+        _;
+    }
+
+    modifier onlyOperator() {
+        if (!hasRole(OPERATOR_ROLE, msg.sender)) revert NotOperator();
+        _;
+    }
+
+    modifier onlyTreasurer() {
+        if (!hasRole(TREASURER_ROLE, msg.sender)) revert NotTreasurer();
+        _;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                        STORAGE RENTAL LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Rent storage for a given fid. The caller must provide an exact payment amount.
+     *
+     * @param fid   The fid that will receive the storage allocation.
+     * @param units Number of storage units to rent.
+     */
+    function rent(uint256 fid, uint256 units) external payable whenNotDeprecated {
+        // Checks
+        if (units == 0) revert InvalidAmount();
+        if (rentedUnits + units > maxUnits) revert ExceedsCapacity();
+        uint256 totalPrice = _price(units);
+        if (msg.value < totalPrice) revert InvalidPayment();
+
+        // Effects
+        rentedUnits += units;
+        emit Rent(msg.sender, fid, units);
+
+        // Interactions
+        if (msg.value > totalPrice) {
```

Comment:
Hmm, I think I'd expect most callers to retrieve the latest `price(units)` and then `rent(fid, units)` with the exact amount returned, rather than add some safety buffer. Allowing overpayments means we gracefully handle the case when the ETH/USD price happened to decrease intra-block or while the tx was waiting. But if the price goes *up* we'll still revert.  

That should be pretty infrequent, at least for users who aren't lowballing gas: to revert, the tx has to be in the same 2-second block as an earlier price feed update that increases the price.

Reply:
Actually, we could probably handle this case too, if we use block number to make the "honor old price" decision: hold on to the previous price, store the block number of the last change, and use the old price if it changed in this block. Then, the price change would only apply starting in the next block, rather than the next tx.